### ISSUE TITLE
Support for external reference for code generation

### ISSUE THREAD

**Issue by jemerald** [swagger-api/swagger-editor#1649]
| Q                                   | A
| ----------------------------------- | -------
| Bug or feature request?             |  Feature
| Which Swagger/OpenAPI version?      | 2.0
| Which Swagger-Editor version?       | 3.2.6
| How did you install Swagger-Editor? | swagger-editor-dist
| Which broswer & version?            | Chrome@latest
| Which operating system?             | Windows 10

### Demonstration API definition

def.yaml

```yaml
swagger: "2.0"
info:
  version: 1.0.0
  title: Swagger Petstore
definitions:
  Pet:
    required:
      - id
      - name
    properties:
      id:
        type: integer
        format: int64
      name:
        type: string
      tag:
        type: string
  Pets:
    type: array
    items:
      $ref: '#/definitions/Pet'
  Error:
    required:
      - code
      - message
    properties:
      code:
        type: integer
        format: int32
      message:
        type: string
```

petstore.yaml

```yaml
swagger: "2.0"
info:
  version: 1.0.0
  title: Swagger Petstore
  license:
    name: MIT
host: petstore.swagger.io
basePath: /v1
schemes:
  - http
consumes:
  - application/json
produces:
  - application/json
paths:
  /pets:
    get:
      summary: List all pets
      operationId: listPets
      tags:
        - pets
      parameters:
        - name: limit
          in: query
          description: How many items to return at one time (max 100)
          required: false
          type: integer
          format: int32
      responses:
        "200":
          description: A paged array of pets
          headers:
            x-next:
              type: string
              description: A link to the next page of responses
          schema:
            $ref: 'def.yaml#/definitions/Pets'
        default:
          description: unexpected error
          schema:
            $ref: 'def.yaml#/definitions/Error'
```

### Configuration (browser query string, constructor, config.yaml)

YAML files are served on HTTP with GET/PUT, editor is opened with the following URL

```
http://localhost:8080/editor/?url=/specifications/petstore.yaml
```

### Expected Behavior

When generating code the models in all the external files should also be generated

### Current Behavior

Code generation doesn't generate any model in `def.yaml`, just the controller classes in `petstore.yaml`.

### Possible Solution

When performing code generation, the editor should traverse the spec for any external references, then load them in and merge into a single spec before sending to the code generation service.

Note: Swagger UI already have similar feature and is able to display the model from `def.yaml` correctly while viewing `petstore.yaml`

**Comment by shockey** [comment]
Hi @jemerald!

This is actually a Swagger-Codegen limitation, so you should open this issue there: https://github.com/swagger-api/swagger-codegen

I'm going to close this out for now, but if the Codegen folks kick it back to Swagger-Editor for some reason, feel free to ask for a reopen.

BTW - thanks for your PRs and issues recently! It's always appreciated 😄

**Comment by jemerald** [comment]
@shockey I'm not sure if it is Swagger-Codegen's problem though...

Correct me if I'm wrong, but as far as I can see, when [performing code generation](https://github.com/swagger-api/swagger-editor/blob/master/src/standalone/topbar/topbar.jsx#L141) from swagger editor, the editor actually calls `POST https://generator.swagger.io/api/gen/servers/{lang}` with the swagger spec in the body. The code gen API does its thing, and respond with a one-off link to download the code in a zip file.

```
      swaggerClient.apis.servers.generateServerForLanguage({
        framework : name,
        body: JSON.stringify({
          spec: specSelectors.specJson() // here
        }),
        headers: JSON.stringify({
          Accept: "application/json"
        })
      })
        .then(res => this.handleResponse(res, { type }))
```

In such an architecture, the codegen service only receives the content from the POST request, and would have no idea about the location of the spec file, nor where to find the externally referenced specs.

The way I see it working, is for the editor to traverse and merge the external references to a single spec before sending it to the codegen service.

What do you think?

**Comment by shockey** [comment]
@jemerald, thanks for the note! I brought this up with the team here, and @webron indicated that the Editor is not meant to do the resolution work for the Codegen.

That being said - there is an opportunity to write a Swagger-Editor plugin that would send the resolved spec to Codegen instead of the unresolved one. If that's something you're interested in doing, let me know!

**Comment by jemerald** [comment]
I don't mind give it a try. But I'm not familiar enough with the whole system so might need some guidance.

Had a brief look though the code in the 3 projects (editor, ui and js) and here are my findings:

- Swagger UI uses the resolver from `swagger-js` to resolve the internal and external `$ref` for display, the resolved spec is stored in `spec.resolved`. All the referenced models gets copied inline.
- Swagger Editor used to send the resolved spec to codegen until [this change](https://github.com/swagger-api/swagger-editor/commit/81445d9054b4aecfb9b8a1ab5557862fae6c8ec3#diff-155f7d148492eda891d2aa058ede1605), and now sends the original spec instead
  - the above change fixed issue #1319 (generated code contains models like `InlineResponse200`), but causes this issue of not supporting external file references

Potential solutions:

1. Revert to use resolved spec for codegen - this is the simplest option but means going back to the `InlineResponse200` problem
1. Option 1, plus rework resolver in `swagger-js` to address the `InlineResponse200` problem (not sure if this is possible as it will be a breaking change in terms of the resolved output)
1. Option 1, plus modify codegen to take the `$$ref` field as hint for model name when generating code
1. Implement new codegen-friendly resolver to be used by editor when generating code

Thoughts?

**Comment by shockey** [comment]
@jemerald, as of now option 1 is a non-starter as we currently have some work underway that does away with the resolved spec altogether, in favor of an on-demand resolver interface.

As for the others - @webron should be able to shed some light on Codegen support for external references (or lack thereof!)
